File: virus.cpp
// University of Central Florida
// 16th Annual High School Programming Tournament
// May 3rd, 2002
//
// Problem Name: Illegal Hawaiian Virus Collection
// Filename: virus.cpp
// Input File: virus.in

#include <iostream>
#include <fstream>
#include <string>

using namespace std;

bool match(string virus, string pattern) {
  int lstart = ((int)virus.size()) - ((int)pattern.size());
  for (int start = 0; start <= lstart; start++) {
    bool ok = true;
    for (int pos = 0; pos < pattern.size(); pos++) {
      if (pattern[pos] == '*') continue;
      if (pattern[pos] != virus[pos+start]) {
        ok = false;
        break;
      }
    }
    if (ok) {
      /*cout << "\n" << virus << "\n";
      for (int i = 0; i < start; i++) cout << ' ';
      cout << pattern << "\n";*/
      return true;
    }
  }
  return false;
}

void main() {
  ifstream fin("virus.in");
  int nSets, nPatterns, nVirii;
  fin >> nSets;

  for (int s = 1; s <= nSets; s++) {
    cout << "Data set #" << s << ":\n";

    string patterns[30];
    fin >> nPatterns;
    for (int p = 0; p < nPatterns; p++) {
      fin >> patterns[p];
    }

    fin >> nVirii;
    for (int v = 1; v <= nVirii; v++) {
      cout << "Virus #" << v << ": ";
      string virus;
      fin >> virus;

      bool matched = false;
      for (int p = 0; p < nPatterns; p++) {
        if (match(virus, patterns[p])) {
          cout << "Nuts. This virus is illegal in Hawaii!\n";
          matched = true;
          break;
        }
      }
      if (!matched) cout << "Cool! Victor can take it with him!\n";
    }

    cout << "\n";
  }
}

File: virus.pas
{
 University of Central Florida
 16th Annual High School Programming Tournament
 May 3rd, 2002

 Problem Name: Illegal Hawaiian Virus Collection
 Filename: virus.pas
 Input File: virus.in
}

program virusChecker;

var
   fin: Text;
   nSets, nVirii, nPatterns: longint;
   s, v, p: longint;
   patterns: ARRAY [ 1 .. 30 ] of String;
   virus: String;
   matched: Boolean;

function match(str, pattern: String):Boolean;

var
   start, pos, lstart: longint;
   ok, retval: Boolean;

begin
   lstart := Length(str) - Length(pattern) + 1;
   retval := false;
   for start := 1 to lstart do begin
      ok := true;
      for pos := 1 to Length(pattern) do begin
         if (pattern[pos] = '*') then continue;
         if (pattern[pos] <> str[pos+start-1]) then begin
            ok := false;
            break;
         end;
      end;
      if ok then begin
        retval := true;
        break;
      end;
    end;
    match := retval;
end;

begin
   assign(fin, 'virus.in');
   reset(fin);

   readln(fin, nSets);

   for s := 1 to nSets do begin
      writeln('Data set #', s, ':');

      readln(fin, nPatterns);
      for p := 1 to nPatterns do
         readln(fin, patterns[p]);

      readln(fin, nVirii);
      for v := 1 to nVirii do begin
         write('Virus #', v, ': ');
         readln(fin, virus);
         matched := False;
         for p := 1 to nPatterns do begin
            if match(virus, patterns[p]) then begin
               matched := True;
               writeln('Nuts. This virus is illegal in Hawaii!');
               break;
            end;
         end;

         if not matched then writeln('Cool! Victor can take it with him!');
      end;

      writeln;

   end;

   close(fin);
end.
